src/pygeometadata/mcf.py

```diff
@@ -0,0 +1,413 @@
+            section (string): the name of a keywords section
+            keywords_type (string): subject matter used to group similar
+                keywords. Must be one of,
+                ('discipline', 'place', 'stratum', 'temporal', 'theme')
+            vocabulary (dict): a dictionary with 'name' and 'url' (optional)
+                keys. Used to describe the source (thesaurus) of keywords
+
+        Raises:
+            TypeError if ``keywords`` is not a list or tuple
+
+        """
+        if not isinstance(keywords, (list, tuple)):
+            raise TypeError(
+                'The first argument of keywords must be a list.'
+                f'received {type(keywords)} instead')
+
+        section_dict = {
+            'keywords': keywords,
+            'keywords_type': keywords_type
+        }
+
+        if vocabulary:
+            section_dict['vocabulary'] = vocabulary
+        self.mcf['identification']['keywords'][section] = section_dict
+
+    def describe_band(self, band_number, name=None, title=None, abstract=None,
+                      units=None):
+        """Define metadata for a raster band.
+
+        Args:
+            band_number (int): a raster band index, starting at 1
+            name (str): name for the raster band
+            title (str): title for the raster band
+            abstract (str): description of the raster band
+            units (str): unit of measurement for the band's pixel values
+        """
+        idx = band_number - 1
+        attribute = self.mcf['content_info']['attributes'][idx]
+        if name is not None:
+            attribute['name'] = name
+        if title is not None:
+            attribute['title'] = title
+        if abstract is not None:
+            attribute['abstract'] = abstract
+        if units is not None:
+            attribute['units'] = units
+
+        self.mcf['content_info']['attributes'].insert(idx, attribute)
+
+    def describe_field(self, name, title=None, abstract=None,
+                       units=None):
+        """Define metadata for a tabular field.
+        Args:
+            name (str): name and unique identifier of the field
+            title (str): title for the field
+            abstract (str): description of the field
+            units (str): unit of measurement for the field's values
+        """
+        def get_attr(attribute_list):
+            for idx, attr in enumerate(attribute_list):
+                if attr['name'] == name:
+                    return idx, attr
+            raise ValueError(
+                f'{self.datasource} has no attribute named {name}')
+
+        idx, attribute = get_attr(self.mcf['content_info']['attributes'])
+
+        if title is not None:
+            attribute['title'] = title
+        if abstract is not None:
+            attribute['abstract'] = abstract
+        if units is not None:
+            attribute['units'] = units
+
+        self.mcf['content_info']['attributes'].insert(idx, attribute)
```

Comment:
Yes...I guess it would be equivalent to `self.mcf['content_info']['attributes'][idx] = attribute`

`attributes` is a list containing `attribute`. `attribute` is being modified in this method, and then being used to replace the pre-existing `attribute` at the same index. Am I missing your meaning?

I added to `test_raster_attributes` and `test_vector_attributes` to better illustrate the behavior I'm going for.

Reply:
My main concern is about the usage of `insert`, which doesn't overwrite the element:

```python
>>> a = [1, 2, 3]
>>> a.insert(2, 'foo')
>>> a
[1, 2, 'foo', 3]
```

We can see this when adding an assertion to the test suite, where we're starting off with a 2-band raster (so 2 `attributes` sections), and the base attribute dict is being copied once per call to `insert`:

```python
>       self.assertEqual(len(mcf.mcf['content_info']['attributes']), 2)
E       AssertionError: 4 != 2
```

Reply:
Ahhh, of course, thanks for spelling it out for me. Patched in e282ca3